### PR TITLE
Fix ESP32 webquery response

### LINKED ISSUE #20102: WebQuery get right answer on 8266 but erratic on ESP32
### PROBLEM DESCRIPTION
I use the same script on 4 tasmota devices : 3 with 8266, 1 with ESP32 which are all querying the same web service every 3 seconds using WebQuery GET. This works perfectly on the 8266 devices, I mean they receives the web service answer on each call; while the ESP32 does not receive the data most of the time, quite erratically.
My web service is correctly called by the device (correct headers and parameter).

It "receives the answer" means it logs like this :
23:49:00.760 RSL: RESULT = {"WebQuery":{"cmd":0, "con":0, "alv":172297}}
23:49:00.771 RSL: RESULT = {"WebQuery":"Done"}

It "does not receive the answer" means it logs like this :
23:46:34.111 RSL: RESULT = {"WebQuery":{"WebQuery":}
23:46:34.120 RSL: RESULT = {"WebQuery":"Done"}

The calling WebQuery is:
=>WebQuery http://www.xxxx.xxxx/connect/v1/device?al=%0nWS% GET [id:%topic%|ip:%lip%]
(The issue is the same in http and https)

Resulting in log:
00:09:07.260 CMD: Grp 0, Cmd 'WEBQUERY', Idx 1, Len 107, Pld -99, Data 'http://www.xxxx.xxxx/connect/v1/device?al=29236 GET [id:01234567890123456789012345678901|ip:192.168.0.38]'

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [ x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ x] Device used (e.g., Sonoff Basic): Rocket Controller ESP32R4
- [x ] Tasmota binary firmware version number used: 13.0.0
  - [ ] Pre-compiled
  - [x ] Self-compiled
- [ x] Flashing tools used: OTA
- [ x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
00:06:47.367 CMD: Backlog Template; Module; GPIO 255
00:06:47.396 RSL: RESULT = {"NAME":"ESP32R4","GPIO":[1,1,576,1,480,1,1,1,1,1,1,1,1,1,1,1,0,640,608,1,0,224,225,1152,0,0,0,0,227,226,32,33,34,0,0,35],"FLAG":0,"BASE":1}
00:06:47.620 RSL: RESULT = {"Module":{"0":"ESP32R4"}}
00:06:47.876 RSL: RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"576":"LedLink_i"},"GPIO3":{"0":"None"},"GPIO4":{"480":"Buzzer"},"GPIO5":{"0":"None"},"GPIO6":{"0":"None"},"GPIO7":{"0":"None"},"GPIO8":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO11":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"},"GPIO18":{"0":"None"},"GPIO19":{"0":"None"},"GPIO20":{"0":"None"},"GPIO21":{"640":"I2C SDA"},"GPIO22":{"608":"I2C SCL"},"GPIO23":{"0":"None"},"GPIO24":{"0":"None"},"GPIO25":{"224":"Relay1"},"GPIO26":{"225":"Relay2"},"GPIO27":{"1152":"RFrecv"},"GPIO32":{"227":"Relay4"},"GPIO33":{"226":"Relay3"},"GPIO34":{"32":"Button1"},"GPIO35":{"33":"Button2"},"GPIO36":{"34":"Button3"},"GPIO37":{"0":"None"},"GPIO38":{"0":"None"},"GPIO39":{"35":"Button4"}}

```
- [N/A ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ X] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
00:08:00.585 CMD: Status 0
00:08:00.590 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota","Tasmota2","Tasmota3","Tasmota4"],"Topic":"TeGOrLS52scMK3fQUN5kolbKfdRVoIul","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
00:08:00.598 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota32/release/tasmota32-safeboot.bin","RestartReason":"Software reset CPU","Uptime":"1T10:58:25","StartupUTC":"2023-11-24T12:09:35","Sleep":50,"CfgHolder":4617,"BootCount":13,"BCResetTime":"2023-11-15T08:02:08","SaveCount":104}}
00:08:00.606 RSL: STATUS2 = {"StatusFWR":{"Version":"13.0.0(TasmoCompiler-esp32generic)","BuildDateTime":"2023-11-24T08:33:51","Core":"2_0_10","SDK":"4.4.5.230614","CpuFrequency":160,"Hardware":"ESP32-D0WD rev.1","CR":"427/699"}}
00:08:00.610 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["VOO-014092",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00206000","00004000","00000000"]}}
00:08:00.619 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":1253,"Free":1856,"Heap":121,"StackLowMark":1,"PsrMax":4093,"PsrFree":4085,"ProgramFlashSize":4096,"FlashSize":16384,"FlashChipId":"1840A1","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","078045CA","000C0001","00000003","00000000","C0000881","00000080","00200000","54000008","00000080"],"Drivers":"1,2,4,8,9,10,12,24,26,29,35,38,50,52,62,68","Sensors":"1,2,127"}}
00:08:00.627 RSL: STATUS5 = {"StatusNET":{"Hostname":"TeGOrLS52scMK3fQUN5kolbKfdRVoIu","IPAddress":"192.168.0.38","Gateway":"192.168.0.1","Subnetmask":"255.255.255.0","DNSServer1":"62.197.111.140","DNSServer2":"109.88.203.3","Mac":"3C:E9:0E:B0:8C:24","IP6Global":"2a02:2788:1b4:cbd:3ee9:eff:feb0:8c24","IP6Local":"fe80::3ee9:eff:feb0:8c24","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
00:08:00.634 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_B08C24","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
00:08:00.639 RSL: STATUS7 = {"StatusTIM":{"UTC":"2023-11-25T23:08:00","Local":"2023-11-26T00:08:00","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"08:15","Sunset":"16:59"}}
00:08:00.648 RSL: STATUS10 = {"StatusSNS":{"Time":"2023-11-26T00:08:00"}}
00:08:00.655 RSL: STATUS11 = {"StatusSTS":{"Time":"2023-11-26T00:08:00","Uptime":"1T10:58:25","UptimeSec":125905,"Heap":120,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":32},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"VOO-014092","BSSId":"50:09:59:F3:80:01","Channel":6,"Mode":"11n","RSSI":100,"Signal":-46,"LinkCount":1,"Downtime":"0T00:00:04"}}}

```
- [x ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
00:09:07.260 CMD: Grp 0, Cmd 'WEBQUERY', Idx 1, Len 107, Pld -99, Data 'http://www.xxxx.xxxx/connect/v1/device?al=29236 GET [id:01234567890123456789012345678901|ip:192.168.0.38]'
00:09:07.461 RSL: RESULT = {"WebQuery":{"WebQuery":}
00:09:07.470 RSL: RESULT = {"WebQuery":"Done"}


```

### TO REPRODUCE
_Steps to reproduce the behavior:_
I can send my script in mp for a real-time test

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
I expect this answer as on 8266 devices:
23:49:00.760 RSL: RESULT = {"WebQuery":{"cmd":0, "con":0, "alv":172297, "cf1":2, "cf2":3, "cf3":4, "cf4":4, "cf5":20}}

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._
I compiled Tasmota 13.0.0 using TasmoCompiler including Script and TLS MQTT support but not Berry.
Here are my user_config_override.h options:
#ifndef ESP32
#define ESP32
#endif

#ifdef USE_BERRY
#undef USE_BERRY
#endif

#ifdef SCRIPT_MAXSSIZE
#undef SCRIPT_MAXSSIZE
#endif
#define SCRIPT_MAXSSIZE 64

#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif

#ifndef USE_WEBSEND_RESPONSE
#define USE_WEBSEND_RESPONSE
#endif

#ifdef USE_RULES
#undef USE_RULES
#endif

#define PRECONFIGURED_SCRIPT ">D 128\n;--- start setup ---\n;......"

#ifndef START_SCRIPT_FROM_BOOT
#define START_SCRIPT_FROM_BOOT
#endif

**(Please, remember to close the issue when the problem has been addressed)**

### PR DESCRIPTION
**Pull request by barbudor**
## Description:

**Related issue (if applicable):** fixes #20102 

Creating a copy of the response seems to solve the problem

## Checklist:
  - [X] The pull request is done against the latest development branch
  - [X] Only relevant files were touched
  - [X] Only one feature/fix was added per PR and the code change compiles without warnings
  - [X] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [X] The code change is tested and works with Tasmota core ESP32 V.2.0.14
  - [X] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

File: tasmota/tasmota_xdrv_driver/xdrv_01_9_webserver.ino [hunk 1]
```diff
@@ -3356,7 +3356,8 @@ int WebQuery(char *buffer) {
         if (http_code == HTTP_CODE_OK || http_code == HTTP_CODE_MOVED_PERMANENTLY) {
 #ifdef USE_WEBSEND_RESPONSE
           // Return received data to the user - Adds 900+ bytes to the code
-          const char* read = http.getString().c_str();  // File found at server - may need lot of ram or trigger out of memory!
+          String response = String(http.getString().c_str());  // File found at server - may need lot of ram or trigger out of memory!
```
Comment (arendst) [inline tasmota/tasmota_xdrv_driver/xdrv_01_9_webserver.ino:3359]: Does it work too with just ``String response = http.getString()``?

Comment (barbudor) [reply tasmota/tasmota_xdrv_driver/xdrv_01_9_webserver.ino]: looks like yes. 15 minutes without error

**Comment by barbudor** [comment]
Stil there is something wrong
My Rule2 doesn't get triggered while it was working in the past:
```
23:42:40.917 RUL: TELE-TIME performs "webquery http://192.168.168.244/myip.php?json GET"
23:42:40.919 SRC: Rule
23:42:40.921 CMD: Grp 0, Cmd 'WEBQUERY', Idx 1, Len 40, Pld -99, Data 'http://192.168.168.244/myip.php?json GET'
23:42:40.959 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":{"myip":"192.168.168.196"}}
23:42:40.973 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":"Done"}
23:42:45.644 WIF: Checking connection...
23:42:50.902 BRY: GC from 3819 to 3400 bytes, objects freed 1/30 (in 1 ms) - slots from 34/76 to 33/76
23:42:50.913 MQT: tele/esp32dev3/STATE = {"Time":"2023-11-26T23:42:50+01:00","Uptime":"0T00:14:16","UptimeSec":856,"Heap":164,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":30},"Wifi":{"AP":1,"SSId":"BNETWLAN","BSSId":"88:C3:97:F2:25:EC","Channel":1,"Mode":"11n","RSSI":100,"Signal":-43,"LinkCount":1,"Downtime":"0T00:00:10"}}
23:42:50.929 RUL: TELE-TIME performs "webquery http://192.168.168.244/myip.php?json GET"
23:42:50.931 SRC: Rule
23:42:50.933 CMD: Grp 0, Cmd 'WEBQUERY', Idx 1, Len 40, Pld -99, Data 'http://192.168.168.244/myip.php?json GET'
23:42:50.975 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":{"myip":"192.168.168.196"}}
23:42:50.988 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":"Done"}
23:43:00.911 BRY: GC from 3819 to 3400 bytes, objects freed 1/30 (in 1 ms) - slots from 34/76 to 33/76
23:43:00.921 MQT: tele/esp32dev3/STATE = {"Time":"2023-11-26T23:43:00+01:00","Uptime":"0T00:14:26","UptimeSec":866,"Heap":164,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":30},"Wifi":{"AP":1,"SSId":"BNETWLAN","BSSId":"88:C3:97:F2:25:EC","Channel":1,"Mode":"11n","RSSI":100,"Signal":-42,"LinkCount":1,"Downtime":"0T00:00:10"}}
23:43:00.938 RUL: TELE-TIME performs "webquery http://192.168.168.244/myip.php?json GET"
23:43:00.940 SRC: Rule
23:43:00.942 CMD: Grp 0, Cmd 'WEBQUERY', Idx 1, Len 40, Pld -99, Data 'http://192.168.168.244/myip.php?json GET'
23:43:00.988 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":{"myip":"192.168.168.196"}}
23:43:01.005 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":"Done"}
23:43:05.674 WIF: Checking connection...
23:43:10.882 BRY: GC from 3852 to 3400 bytes, objects freed 1/30 (in 1 ms) - slots from 34/76 to 33/76
23:43:10.893 MQT: tele/esp32dev3/STATE = {"Time":"2023-11-26T23:43:10+01:00","Uptime":"0T00:14:36","UptimeSec":876,"Heap":162,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":30},"Wifi":{"AP":1,"SSId":"BNETWLAN","BSSId":"88:C3:97:F2:25:EC","Channel":1,"Mode":"11n","RSSI":100,"Signal":-43,"LinkCount":1,"Downtime":"0T00:00:10"}}
23:43:10.907 RUL: TELE-TIME performs "webquery http://192.168.168.244/myip.php?json GET"
23:43:10.909 SRC: Rule
23:43:10.910 CMD: Grp 0, Cmd 'WEBQUERY', Idx 1, Len 40, Pld -99, Data 'http://192.168.168.244/myip.php?json GET'
23:43:10.947 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":{"myip":"192.168.168.196"}}
23:43:10.962 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":"Done"}
23:43:20.887 BRY: GC from 3819 to 3400 bytes, objects freed 1/30 (in 1 ms) - slots from 34/76 to 33/76
23:43:20.898 MQT: tele/esp32dev3/STATE = {"Time":"2023-11-26T23:43:20+01:00","Uptime":"0T00:14:46","UptimeSec":886,"Heap":164,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":30},"Wifi":{"AP":1,"SSId":"BNETWLAN","BSSId":"88:C3:97:F2:25:EC","Channel":1,"Mode":"11n","RSSI":100,"Signal":-44,"LinkCount":1,"Downtime":"0T00:00:10"}}
23:43:20.912 RUL: TELE-TIME performs "webquery http://192.168.168.244/myip.php?json GET"
23:43:20.913 SRC: Rule
23:43:20.915 CMD: Grp 0, Cmd 'WEBQUERY', Idx 1, Len 40, Pld -99, Data 'http://192.168.168.244/myip.php?json GET'
23:43:20.953 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":{"myip":"192.168.168.196"}}
23:43:20.971 MQT: stat/esp32dev3/WEBQUERY = {"WebQuery":"Done"}
```

**Comment by barbudor** [comment]
I need to check further because it seems that rules get impacted
rules on webquery# that were working on esp8266 doesn't work anymore

**Comment by arendst** [comment]
I'll do some chks too using this on my server:
```
<?php
    $ip = getenv("REMOTE_ADDR");

    if (isset($_GET["json"])) {
        echo "{myip:\"$ip\"}";
    } else {
        echo "myip = $ip";
    }
?>
```
It works on ESP8266 but fails on ESP32 and ESP32-C3.

**Comment by arendst** [comment]
Your PR fixes at least the missing/corrupted response so I'll merge this.

The issue with rules executing while another rule is busy (Rules.busy flag) is a different longstanding problem without a sinple solution.

**Comment by sfromis** [comment]
Doing a simple WebQuery (3-character response) started to work now, on an ESP32-C3.

**Comment by barbudor** [comment]
strange because characters are copied 1 by 1 from the http response to the MQTT message
I will try to investigate tonight

**Comment by barbudor** [comment]
> UPDATE1: Your rule2 needs to be changed to: rule2 on webquery#myip do var1 %value% endon

That what I was first thinking but I found old notes from rules I was using more than a year ago that were apparently working like that. May be a change in Tasmota. 
But webquery#myip makes much more sense

> UPDATE2: The rules works fine when executed manually but corrupt the "Done" result.

So that seems to confirm that there is something broken in the kingdom of ext_vsnprinf_P ?

**Comment by arendst** [comment]
> So that seems to confirm that there is something broken in the kingdom of ext_vsnprinf_P ?

No, I don't think something is off with the long standing ext_vsnprintf_P. Something is off with releasing Strings. ANyway, I'll remove the unnneeded "Done" message resulting in a single response as expected.

> However, I'm still seeing WebQuery ESP32 badness with a file containing {"Level":"47%"}:

Works fine on ESP8266 (but removes the %) and fails on any ESP32. Investigating too with focus on ``HttpClientLight.cpp`` not used by ESP8266....

**Comment by arendst** [comment]
Should be fixed now.

**Comment by arendst** [comment]
> Could/can anything useful be done with % formatting around WebQuery or other "non-intuitive" places?

Not anymore ;-) All formatting is now disabled in ``WebQuery`` response.

Using the ``Done`` for a rule trigger is useless as the previous webquery response is overwritten by the ``Done`` message.

**Pull request merged**